### PR TITLE
Push downstream models to single Hub repository

### PR DESCRIPTION
**Pull request by lewtun**
This PR refactors the `Runner.push_to_hf_hub` method to push the training metrics of each fine-tuning experiment to a single repository on the Hugging Face Hub.  

The logic is to create a repository with the naming convention:

```
{hf_hub_user_or_org}__{upstream_model_name}__{upstream_model_commit_sha}
```

See [here](https://huggingface.co/superb/superb-test-org__test-submission-with-weights__2323d47e588aa02648ac1770568eeaa203431535) for an example where the model repository structure is of the form:

```
model_repo
├── asr
│   ├── args_2021-09-23-10-51-44.yaml
│   ├── char.dict
│   ├── config_2021-09-23-10-51-44.yaml
│   ├── events.out.tfevents.1632387113.vorace
│   ├── model.ckpt
│   ├── model.py
│   └── README.md
└── diarization
    ├── args_2021-09-23-11-01-28.yaml
    ├── config_2021-09-23-11-01-28.yaml
    ├── events.out.tfevents.1632387697.vorace
    ├── model.ckpt
    ├── model.py
    ├── README.md
    ├── scoring
    └── upstream_rate
```

See [here](https://huggingface.co/superb/hubert__508944ac) for an example using an upstream models from the PyTorch Hub.

## TODO

- [x] Refactor to support upstream models from PyTorch Hub (which don't have commit SHA)
- [x] Tidy up templates (unlikely to be needed in the current approach)
- [x] Add README template in root of model repository
- [x] Test this also works with the `example` expert template for the upstream model

### CODE REVIEW

File: s3prl/downstream/runner.py [hunk 1]
```diff
@@ -448,63 +448,63 @@ def inference(self):
...
+        upstream_model_id = model_info.modelId.replace("/", "__")
+        repo_name = f"{upstream_model_id}__{commit_sha}"
+        # Create downstream repo on the Hub
         repo_url = HfApi().create_repo(
             token=huggingface_token,
             name=repo_name,
             organization=organization,
             exist_ok=True,
-            private=True,
+            private=False,
```
Comment (lewtun) [inline s3prl/downstream/runner.py:471]: As discussed offline with @leo19941227 we now set the fine-tuned model repos to be public by default so the SUPERB participants can inspect the tensorboard logs etc

File: s3prl/downstream/runner.py [hunk 2]
```diff
@@ -29,6 +29,49 @@
 
 SAMPLE_RATE = 16000
 
+MODEL_CARD_MARKDOWN = """---
+datasets:
```
Comment (lewtun) [inline s3prl/downstream/runner.py:33]: These YAML tags are for discoverability on the Hugging Face Hub

File: s3prl/downstream/runner.py [hunk 3]
```diff
@@ -448,64 +496,76 @@ def inference(self):
...
         if TEMPLATES_PATH.exists():
-            shutil.copytree(TEMPLATES_PATH, REPO_PATH, dirs_exist_ok=True)
+            shutil.copytree(TEMPLATES_PATH, REPO_TASK_DIR, dirs_exist_ok=True)
         else:
             print(f"[Runner] - No Hugging Face Hub template found for downstream task! Experiment files will still be pushed to the Hub in raw form")
 
         # Copy checkpoints, tensorboard logs, and args / configs
-        shutil.copytree(self.args.expdir, REPO_PATH, dirs_exist_ok=True, ignore=shutil.ignore_patterns(HF_HUB_DIR))
-
-        # Inject upstream model name into model card
```
Comment (lewtun) [inline s3prl/downstream/runner.py:485]: I removed this step since I don't think it is need any longer

**Comment by lewtun** [comment]
I think this is now ready for a review @leo19941227 🥳 !

**Comment by leo19941227** [comment]
> I think this is now ready for a review @leo19941227 🥳 !

Thanks @lewtun ! I have two comments:

- As we discussed, we might want to disentangle the `push_to_hub` with the `Runner.train`. Does this functionality also enabled by this PR?
- Another factor we not yet discussed is that, I find that each upstream might have various `expdirs` for each task. That is, at least we will do the learning rate search. So I think it might be nice to organize the REPO in the following structure:

```
model_repo
├── asr
│    ├─ asr_lr1e-4 (args.expdir)
│    │   ├── args_2021-09-23-10-51-44.yaml
│    │   ├── char.dict
│    │   ├── config_2021-09-23-10-51-44.yaml
│    │   ├── events.out.tfevents.1632387113.vorace
│    │   ├── model.ckpt
│    │   ├── model.py
│    │   └── README.md
│    └─ asr_lr1e-4 (args.expdir)
│         ├── args_2021-09-23-10-51-44.yaml
│         ├── char.dict
│         ├── config_2021-09-23-10-51-44.yaml
│         ├── events.out.tfevents.1632387113.vorace
│         ├── model.ckpt
│         ├── model.py
│         └── README.md
│
│
```

**Comment by lewtun** [comment]
Hey @leo19941227 I've tweaked the save step so that we now create one subdirectory for every new `args.expname` - this should cover the case mentioned in your second point and you can find an example repo [here](https://huggingface.co/superb/superb-test-org__test-submission-with-example-expert__d609b3c32044e50e3d5e9067bd97af1b42f04b0e/tree/main/asr).

As discussed offline, it should be possible to implement a generic `push_to_`hub` function for the post-evaluation step using `huggingface_hub.Repository.push_to_hub` - I'll leave that to you to implement :)

File: s3prl/downstream/runner.py [hunk 4]
```diff
@@ -448,64 +496,76 @@ def inference(self):
...
-            private=True,
+            private=False,
         )
         print(f"[Runner] - Created Hub repo: {repo_url}")
 
         # Download repo and copy templates
         HF_HUB_DIR = "hf_hub"
-        REPO_PATH = os.path.join(self.args.expdir, HF_HUB_DIR, repo_name)
+        REPO_ROOT_DIR = os.path.join(self.args.expdir, HF_HUB_DIR, repo_name)
+        REPO_TASK_DIR = os.path.join(REPO_ROOT_DIR, self.args.downstream, self.args.expname)
```
Comment (lewtun) [inline s3prl/downstream/runner.py:530]: Here is where I've added the experiment name to the path we store all the training artifacts

Comment (leo19941227) [reply s3prl/downstream/runner.py:530]: Yes this looks good to me!

File: s3prl/downstream/runner.py (hunk 1 above)
Comment (leo19941227) [reply s3prl/downstream/runner.py:523]: Sure LGTM!

File: s3prl/downstream/runner.py (hunk 3 above)
Comment (leo19941227) [reply s3prl/downstream/runner.py:485]: OK!

**Review by leo19941227 — COMMENTED** [review]
Thanks @lewtun !
The PR looks good to me. Do you think it is ready to merge or are there some other changes you are now making? (Since I saw some action items not yet checked on the list in the first message.

**Comment by lewtun** [comment]
Thanks for the review @leo19941227 ❤️ ! I've just addressed the last item (removing the downstream templates) so this should be good to go!

**Comment by leo19941227** [comment]
Cool! I am merging the PR. Thanks @lewtun !

**Review by leo19941227 — APPROVED** [review]
